specs/networking/rpc-interface.md

```diff
@@ -0,0 +1,292 @@
+
+## Messages
+
+### Hello
+
+**Method ID:** `0`
+
+**Body**:
+
+```
+(
+    network_id: uint8
+    chain_id: uint8
+    latest_finalized_root: bytes32
+    latest_finalized_epoch: uint64
+    best_root: bytes32
+    best_slot: uint64
+)
+```
+
+Clients exchange `hello` messages upon connection, forming a two-phase handshake. The first message the initiating client sends MUST be the `hello` message. In response, the receiving client MUST respond with its own `hello` message.
+
+Clients SHOULD immediately disconnect from one another following the handshake above under the following conditions:
+
+1. If `network_id` belongs to a different chain, since the client definitionally cannot sync with this client.
+2. If the `latest_finalized_root` shared by the peer is not in the client's chain at the expected epoch. For example, if Peer 1 in the diagram below has `(root, epoch)` of `(A, 5)` and Peer 2 has `(B, 3)`, Peer 1 would disconnect because it knows that `B` is not the root in their chain at epoch 3:
+
+```
+              Root A
+
+              +---+
+              |xxx|  +----+ Epoch 5
+              +-+-+
+                ^
+                |
+              +-+-+
+              |   |  +----+ Epoch 4
+              +-+-+
+Root B          ^
+                |
++---+         +-+-+
+|xxx+<---+--->+   |  +----+ Epoch 3
++---+    |    +---+
+         |
+       +-+-+
+       |   |  +-----------+ Epoch 2
+       +-+-+
+         ^
+         |
+       +-+-+
+       |   |  +-----------+ Epoch 1
+       +---+
+```
+
+Once the handshake completes, the client with the higher `latest_finalized_epoch` or `best_slot` (if the clients have equal `latest_finalized_epoch`s) SHOULD request beacon block roots from its counterparty via `beacon_block_roots` (i.e., RPC method `10`).
+
+### Goodbye
+
+**Method ID:** `1`
+
+**Body:**
+
+```
+(
+    reason: uint64
+)
+```
+
+Client MAY send `goodbye` messages upon disconnection. The reason field MAY be one of the following values:
+
+- `1`: Client shut down.
+- `2`: Irrelevant network.
+- `3`: Fault/error.
+
+Clients MAY define custom goodbye reasons as long as the value is larger than `1000`.
+
+### Get Status
+
+**Method ID:** `2`
+
+**Request Body:**
+
+```
+(
+	sha: bytes32
+	user_agent: bytes
+	timestamp: uint64
+)
+```
+
+**Response Body:**
+
+```
+(
+	sha: bytes32
+	user_agent: bytes
+	timestamp: uint64
+)
+```
+
+Returns metadata about the remote node.
+
+### Request Beacon Block Roots
+
+**Method ID:** `10`
+
+**Request Body**
+
+```
+(
+	start_slot: uint64
+	count: uint64
+)
+```
+
+**Response Body:**
+
+```
+# BlockRootSlot
+(
+    block_root: bytes32
+    slot: uint64
+)
+
+(
+    roots: []BlockRootSlot
+)
+```
+
+Requests a list of block roots and slots from the peer. The `count` parameter MUST be less than or equal to `32768`. The slots MUST be returned in ascending slot order.
```

Comment:
I have some questions regarding the response:

- Can you skip a slot (e.g., `1, 3, 4`)?
- Can you return less/more than the requested roots?
- Can you start at higher slot?
- Can you return a slot outside of the `start_slot + count` bounds?

Maybe _"The slots MUST be returned in ascending slot order."_ is succinct already? If this is the case we could add something like "The only requirements for `roots` are ...".

P.S. this is my first comment, thanks for making this doc!